src/devices/bus/epson_qx/option.h

```diff
@@ -0,0 +1,276 @@
+	option_slot_device(const machine_config &mconfig, const char *tag, device_t *owner, T &&bus_tag, int slot, U &&opts, const char *dflt)
+		: option_slot_device(mconfig, tag, owner, 0)
+	{
+		option_reset();
+		opts(*this);
+		set_default_option(dflt);
+		set_fixed(false);
+		m_bus.set_tag(std::forward<T>(bus_tag));
+		m_slot = slot;
+	}
+	option_slot_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	DECLARE_WRITE_LINE_MEMBER(inth1_w);
+	DECLARE_WRITE_LINE_MEMBER(inth2_w);
+	DECLARE_WRITE_LINE_MEMBER(intl_w);
+
+	DECLARE_WRITE_LINE_MEMBER(drqf_w);
+	DECLARE_WRITE_LINE_MEMBER(drqs_w);
+
+	DECLARE_WRITE_LINE_MEMBER(rdyf_w);
+	DECLARE_WRITE_LINE_MEMBER(rdys_w);
+
+	DECLARE_WRITE_LINE_MEMBER(eopf);
+	DECLARE_WRITE_LINE_MEMBER(eops);
+
+	auto dmas_w_callback() { m_dmas_w_cb.bind(); }
+	auto dmas_r_callback() { m_dmas_r_cb.bind(); }
+	auto dmaf_w_callback() { m_dmaf_w_cb.bind(); }
+	auto dmaf_r_callback() { m_dmaf_r_cb.bind(); }
+
+	auto in_eopf_callback() { return m_eopf_cb.bind(); }
+	auto in_eops_callback() { return m_eops_cb.bind(); }
+protected:
+	// device-level overrides
+	virtual void device_start() override;
+
+	// configuration
+	required_device<option_bus_device> m_bus;
+	devcb_write8 m_dmas_w_cb;
+	devcb_read8  m_dmas_r_cb;
+	devcb_write8 m_dmaf_w_cb;
+	devcb_read8  m_dmaf_r_cb;
+
+	devcb_write_line m_eopf_cb;
+	devcb_write_line m_eops_cb;;
+
+	int m_slot;
+};
+
+// device type definition
+DECLARE_DEVICE_TYPE(OPTION_BUS_SLOT, option_slot_device)
+
+/* Epson Bus Device */
+
+class option_bus_device :  public device_t
+{
+public:
+	// construction/destruction
+	option_bus_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	template <typename T> void set_iospace(T &&tag, int spacenum) { m_iospace.set_tag(std::forward<T>(tag), spacenum); }
+	void set_memview(memory_view::memory_view_entry &view) { m_view = &view; }
+
+	auto out_inth1_callback() { return m_inth1_cb.bind(); }
+	auto out_inth2_callback() { return m_inth2_cb.bind(); }
+	template <unsigned Slot> auto out_intl_callback() { return m_intl_cb[Slot].bind(); }
+
+	auto out_drqf_callback() { return m_drqf_cb.bind(); }
+	template <unsigned Slot> auto out_drqs_callback() { return m_drqs_cb[Slot].bind(); }
+
+	auto out_rdyf_callback() { return m_rdyf_cb.bind(); }
+	auto out_rdys_callback() { return m_rdys_cb.bind(); }
+
+	memory_view::memory_view_entry& memview() const { return *m_view; }
+	address_space &iospace() const { return *m_iospace; }
+
+	void dackf_w(uint8_t data);
+	uint8_t dackf_r();
+	template <unsigned Slot> void dacks_w(uint8_t data) {  (*this)[Slot]->m_dmas_w_cb(data);  }
+	template <unsigned Slot> uint8_t dacks_r() { return (*this)[Slot]->m_dmas_r_cb(); }
+
+	template <void (option_slot_device::*eop_callback)(int)> void eop(int state) {
+		for (int i = 0; i < m_slot_list.size(); ++i) {
+			option_slot_device *slot = m_slot_list[i];
+			(slot->*eop_callback)(state);
+		}
+	}
```

Comment:
This essentially applies a “write line” call to all the slots, it doesn’t seem to actually be tied to the EOP functions at all.  Could it be named better?  Also, it could use a range-based `for` loop.